transaction.go

```diff
@@ -162,313 +191,512 @@ func (t *Transaction) UnmarshalJSON(bytes []byte) error {
 	}
 
+	svCtx.WorkingSet.UnlockedIdents.AddInputUnlockedBy(ownerIdent.Key(), inputIndex)
+
 	return nil
 }
 
-// SemanticallyValidateInputs checks that every referenced UTXO is available, computes the input sum
-// and returns functions which can be called to verify the signatures.
-// This function should only be called from SemanticallyValidate().
-func (t *Transaction) SemanticallyValidateInputs(utxos InputToOutputMapping, transaction *TransactionEssence, txEssenceBytes []byte) (uint64, []SigValidationFunc, error) {
-	var sigValidFuncs []SigValidationFunc
-	var inputSum uint64
-	seenInputAddr := make(map[string]int)
+// TxSemanticOutputsSender validates that for SenderFeatureBlock occurring on the output side,
+// the given identity is unlocked on the input side.
+func TxSemanticOutputsSender() TxSemanticValidationFunc {
+	return func(svCtx *SemanticValidationContext) error {
+		for outputIndex, output := range svCtx.WorkingSet.Tx.Essence.Outputs {
+			senderFeatureBlock := output.FeatureBlocks().MustSet().SenderFeatureBlock()
+			if senderFeatureBlock == nil {
+				continue
+			}
 
-	for i, input := range transaction.Inputs {
-		in, alreadySeen := input.(*UTXOInput)
-		if !alreadySeen {
-			return 0, nil, fmt.Errorf("%w: unsupported input type at index %d", ErrUnknownInputType, i)
+			// check unlocked
+			sender := senderFeatureBlock.Address
+			if _, isUnlocked := svCtx.WorkingSet.UnlockedIdents[sender.Key()]; !isUnlocked {
+				return fmt.Errorf("%w: output %d", ErrSenderFeatureBlockNotUnlocked, outputIndex)
+			}
 		}
+		return nil
+	}
+}
 
-		// check that we got the needed UTXO
-		utxoID := in.ID()
-		utxo, has := utxos[utxoID]
-		if !has {
-			return 0, nil, fmt.Errorf("%w: UTXO for ID %v is not provided (input at index %d)", ErrMissingUTXO, utxoID, i)
-		}
+// TxSemanticDeposit validates that the IOTA tokens are balanced from the input/output side.
+// It additionally also incorporates the check whether return amounts via StorageDepositReturnUnlockCondition(s) for specified identities
+// are fulfilled from the output side.
+func TxSemanticDeposit() TxSemanticValidationFunc {
+	return func(svCtx *SemanticValidationContext) error {
+		// note that due to syntactic validation of outputs, input and output deposit sums
+		// are always within bounds of the total token supply
+		var in, out uint64
+		inputSumReturnAmountPerIdent := make(map[string]uint64)
+		for inputID, input := range svCtx.WorkingSet.InputSet {
+			in += input.Deposit()
+
+			unlockCondSet := input.UnlockConditions().MustSet()
+			storageDepositReturnUnlockCondition := unlockCondSet.StorageDepositReturn()
+			if storageDepositReturnUnlockCondition == nil {
+				continue
+			}
 
-		var err error
-		deposit, err := utxo.Deposit()
-		if err != nil {
-			return 0, nil, fmt.Errorf("unable to get deposit from UTXO %v (input at index %d): %w", utxoID, i, err)
-		}
-		inputSum += deposit
+			if !unlockCondSet.HasExpirationCondition() {
+				continue
+			}
```

Comment:
Which case does this check implement?

Reply:
Removed in [removes a short circuit check](https://github.com/iotaledger/iota.go/pull/295/commits/ab477d4e698da1d92633628aff1f1bfedd1fd00e)